doc: add introduction sentence for CJS (#42491)
nodejs/node: doc/api/modules.md

<!--name=module-->

CommonJS modules are the classic format to package JavaScript code for reuse. If

> you are interested in writing browser-compatible code, check out the

Reviewer: I think this is disingenuous; ES Modules aren’t browser-compatible unless they omit all non-shimmable core modules, node globals, and also use no bare specifiers (since import maps aren’t a browser standard yet).

Both module systems are almost identically as browser-compatible, with similar caveats.

Reviewer 2: > I think this is disingenuous

I'm guessing you didn't mean it this way, but being "disingenuous" involves intending to mislead.  Let's stick to comments specifically about the text. "I think this is incorrect" communicates the important part without suggesting the author is trying to be deceptive.

Reviewer: Oh, my apologies. I meant to convey that it will lead the reader to the wrong conclusions.

Author: @ljharb I tried to chose my words carefully, but maybe not enough 😅 what I meant to say if the user is trying to author code that should work on the browser, they should probably be using ESM. Im not sure it’s « wrong », but maybe that’s just an opinion we don’t want to have in our docs. Are you suggesting we leave that part out, or do you have a suggestion to improve the phrasing and remove ambiguity?

Reviewer: I think that's an opinion, yes, and one that doesn't actually hold as well as "you should write CJS if you want to target the browser" :-) but we shouldn't need to debate that here.

I'm suggesting that be left out, since it isn't meaningfully better or different regardless of which module format you author, and to be precise, we'd have to list a ton of caveats - for both module systems.

The docs should simply describe the two module systems, and avoid suggesting which one users should use and why.

[ECMAScript modules][] documentation.

> In the CommonJS module system, each file is treated as a separate module. For

Reviewer: This is also true in ES Modules, it’s just that files are addresses as URLs.

Author: Yep, note that this sentence was there before – and probably before ESM was implemented in Node.js. Are you saying I should remove it, or rephrase it, or were you simply pointing out a fact?

Reviewer: I'd just say, "in node, each file is treated as a separate module" such that it applies to both module systems. This will remain true even if we add HTTPS imports, and even if the language adds some form of module blocks.

<!--name=module-->

> CommonJS modules are the classic format to package JavaScript code for Node.js.

Reviewer 3: ```suggestion
CommonJS modules are the original way to package JavaScript code for Node.js.
```

Words like “classic” or “legacy” imply deprecation.

> 

Reviewer 3: ```suggestion
Node.js also supports the cross-platform standard [ECMAScript modules][].

```

I think it’s important to call out and link to the ESM docs here, and it’s good to imply that they’re more a compatible choice. We don’t need to dive into the details of which platforms they’re compatible with and what needs to be done to write cross-platform code, but we also don’t need to avoid all mention that ESM offers more compatibility.

Reviewer 2: I know there's a lot of nuance and history here so maybe I'm misguided on this--and if so, please feel totally free to ignore me--but I'd prefer avoiding "cross-platform" because it has always implied operating systems to me. Maybe "Node.js also supports [ECMAScript modules][] which are also supported in browsers." or something similar to that?

Reviewer 3: ESM is also supported in Deno, and Cloudflare Workers, to name just two non-browser environments. I’d rather not restrict this to “Node and browsers” which implies that Node is the only non-browser platform of significance. It might even be _more_ relevant that people want to write cross-platform server-side code, rather than cross-platform full-stack code.

The structure of the sentence is such that the other platforms that support ESM need to be things that are comparable to Node; the sentence says that Node is one of the supporters, so therefore the other supporters (the other platforms) need to be things that fall in the same category as Node. So even if someone initially thinks of operating systems as being platforms, the sentence structure should clue them in that since Node isn’t an operating system, the “platforms” in question here need to be other JavaScript runtimes like Node.

Author: ```suggestion
Node.js also supports the standard [ECMAScript modules][], which allows to write more
portable code.

```
```suggestion
Node.js also supports the cross-ecosystem standard [ECMAScript modules][].

```
wdyt?

Reviewer 4: Why not just "the standard" ?

Reviewer: "the language standard"? there's far more than one "standard", and CJS is also a standard, just a node-specific one.

Reviewer 4: To be clear, my suggestion was to remove "cross-ecosystem" from the sentence and leave "the standard ECMAScript modules".

Reviewer 3: The cross-platform aspect is the point.

Reviewer 4: I don't think it's necessary to emphasize this point.

Reviewer 3: > I don’t think it’s necessary to emphasize this point.

So you’re proposing “Node.js also supports standard ECMAScript modules.” instead of “Node.js also supports the cross-platform standard ECMAScript modules.”? Personally I find the former to be so uninformative as to be almost suspect. As a reader new to this, I’d read it and be like “. . . and? Why does Node support two systems? Is the lack of any information about ECMAScript modules meant to imply that I shouldn’t use it? Or is there a certain condition under which it’s preferable, and if so, what?”

The primary reason to use ESM is because it’s the cross-platform, cross-runtime, official standard. That’s a big reason to prefer it over CommonJS. We _do_ get people posting issues on here asking us about which system they should use, since Node has been taking an officially neutral stance that both module systems are first-class and fully supported. The answer is that they can use whichever they prefer, and they might prefer ESM if they want their code to be potentially runnable in places other than Node. Saying so in the docs would help alleviate some confusion.

Reviewer: runnable in not-node *without a build process*, is the important caveat i think helps - because CJS is always runnable anywhere, with a build process.

Reviewer 5: Tbh, I think this sentence can be added in a separate PR because this PR is about adding an introduction sentence about CJS, not ESM.

Author: I'd prefer to use `cross-runtime` over `cross-platform`: cross-platform is a given in Node.js (any JS you write for Node.js would work on all supported platforms except a few documented exceptions). Cross-runtime is what would be interesting for some users, but also I agree with Jordan that it's more of an opinion or a bet on future state of things at this point – and I'd say it's OK to have opinionated docs as long as those opinions have consensus, but that doesn't seem to be the case here.

What do you think on the following (maybe less informative, but more likely to have consensus?)

```markdown
CommonJS modules are the classic format to package JavaScript code for Node.js.
It's not to be confused with standard [ECMAScript modules][].
```

Reviewer 3: “It’s not to be confused” feels weird; was there much confusion between the two? And I think it’s important to mention the compatibility reason that someone might want to look at ESM.

```
CommonJS modules are the original way to package JavaScript code for Node.js.
Node.js also supports the [ECMAScript modules][] standard supported by
browsers and other JavaScript runtimes.
```
